resources/type-system/inference.md

```diff
@@ -970,6 +970,391 @@ with respect to `L` under constraints `C0`
+expression artifacts rather than the expressions themselves._
+
+## Soundness guarantees
+
+Along with the notion of _expression soundness_ described above, there are
+several other soundness guarantees made by Dart:
+
+- It is guaranteed that if a method, function, or constructor having return type
+  `T` is invoked, and the method completes with a value `v`, then `v` will be an
+  instance of `T` (with appropriate type substitutions, in the case of a generic
+  method). This guarantee is known as _return value soundness_.
+
+- It is guaranteed that if a method having static type `T` is torn off, then the
+  resulting value will be an instance of `T`. This guarantee is known as
+  _tear-off soundness_.
+
+- It is guaranteed that if a future is an instance of the type `Future<T>`, and
+  it completes with a value `v`, then `v` will be an instance of `T`. This
+  guarantee is known as _future soundness_.
+
+The rules below include informal sketches of a proof that each of the above
+soundness guarantees holds. These are non-normative, so they are typeset in
+_italics_.
+
+# Coercions
+
+_Coercion_ is a type inference step that transforms an expression artifact `m'`,
+and a desired static type `T`, into an expression artifact `m` whose static type
+is `T`.
+
+_Coercions are used in most situations where the existing spec calls for an
+assignability check._
+
+Coercion of an expression artifact `m'` to type `T` produces an expression
+artifact `m` with static type `T`, where `m` is determined as follows:
+
+- Let `T'` be the static type of `m'`.
+
+- Define `m` as follows:
+
+  - If `T' <: T`, then let `m` be an expression artifact whose runtime behavior
+    is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - `m` completes with the value `o`. _Expression soundenss follows from the
+      fact that since `T' <: T`, `o` must also be an instance of the type `T`._
+
+  - Otherwise, if `T'` is `dynamic`, then let `m` be an expression artifact
+    whose runtime behavior is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - If `o` is an instance of the type `T`, `m` completes with the value
+      `o`. _Expression soundness follows trivially._
+
+    - Otherwise, `m` completes with an exception. _Expression soundness follows
+      trivially._
+
+  - Otherwise, if `T'` is an interface type that contains a method called `call`
+    with type `U`, and `U <: T`, then let `m` be an expression artifact whose
+    runtime behavior is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - Let `o'` be the result of tearing off the `call` method of `o`. _This is
```

Comment:
For what it's worth, I've never been a big fan of using `'` in notation since it's easy to overlook. It took me a few reads to correctly track where the text was using `o` and where `o'`.

Maybe `original` and `coerced`? Or `o` and `oc` if those are too long?

Reply:
Yeah, that's a good point. I've removed the `'`s from this section (and other sections). Elsewhere in the doc I'm using numeric suffixes to distinguish things (`T_1`, `T_2`, etc.) so I switched to that convention here as well.